File: NeuroflowLib/layer_connections.cpp
#include "stdafx.h"
#include "layer_connections.h"
#include "layer.h"

USING

layer_connections::layer_connections(layer_connections::connection_type type) :
_type(type)
{
}

bool layer_connections::must_init() const
{
    return !_connectedLayer;
}

void layer_connections::init(const layer_ptr& connectedLayer)
{
    assert(!_connectedLayer);
    assert(connectedLayer);
    _connectedLayer = connectedLayer;
}

void layer_connections::add_one_way(const layer_ptr& layer)
{
    add(layer, flow_direction::one_way);
}

void layer_connections::add_two_way(const layer_ptr& layer)
{
    add(layer, flow_direction::two_way);
}

void layer_connections::add_one_way_to_source(const layer_ptr& layer)
{
    add(layer, flow_direction::one_way_to_source);
}

void layer_connections::add(const layer_ptr& layer, flow_direction direction)
{
    verify_arg(layer != null, "layer");
    verify_arg(direction == flow_direction::one_way || direction == flow_direction::one_way_to_source || direction == flow_direction::two_way, "direction");

    if (find_if(_otherLayers.cbegin(), _otherLayers.cend(), [&](_citem_t(_otherLayers) p) { return p.second == layer; }) == _otherLayers.cend())
    {
        with_other_side_update_suppressed(layer,
        [&]()
        {
            if (_type == connection_type::output)
            {
                layer->input_connections().add(_connectedLayer, direction);
            }
            else
            {
                layer->output_connections().add(_connectedLayer, direction);
            }
        });
        _otherLayers.emplace_back(direction, layer);
    }
}

bool layer_connections::remove(const layer_ptr& layer)
{
    verify_arg(layer != null, "item");

    auto toRemove = find_if(_otherLayers.cbegin(), _otherLayers.cend(), [&](_citem_t(_otherLayers) p) { return p.second == layer; });
    if (toRemove != _otherLayers.cend())
    {
        _otherLayers.erase(toRemove);
        with_other_side_update_suppressed(layer,
        [&]()
        {
            if (_type == connection_type::output)
            {
                layer->input_connections().remove(_connectedLayer);
            }
            else
            {
                layer->output_connections().remove(_connectedLayer);
            }
        });
        return true;
    }
    return false;
}

void layer_connections::clear()
{
    for (auto& layer : _otherLayers)
    {
        if (_type == connection_type::output)
        {
            layer.second->input_connections().remove(_connectedLayer);
        }
        else
        {
            layer.second->output_connections().remove(_connectedLayer);
        }
    }
    _otherLayers.clear();
}

layers_t layer_connections::connected_layers(flow_direction direction) const
{
    return _otherLayers 
        | where([=](const other_layer_t& layer) { return int(layer.first & direction) != 0; })
        | select([](const other_layer_t& layer) { return layer.second; });
}

void layer_connections::with_other_side_update_suppressed(const layer_ptr& layer, const std::function<void()>& method)
{
    if (_suppressOtherSideUpdate)
    {
        layer->input_connections()._suppressOtherSideUpdate = true;
        layer->output_connections()._suppressOtherSideUpdate = true;
        try
        {
            finally f([&]()
            {
                layer->input_connections()._suppressOtherSideUpdate = false;
                layer->output_connections()._suppressOtherSideUpdate = false;
            });

            method();
        }
        catch (...)
        {
            throw;
        }
    }
}

File: _vault/NFBak/Neuroflow_vsonline/Neuroflow/Neuroflow-Reborn/Neuroflow/NeuralNetworks/LayerConnections.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Neuroflow.NeuralNetworks
{
    public sealed class LayerConnections
    {
        internal enum ConnectionType { Input, Output }

        internal LayerConnections(ConnectionType type, Layer connectedLayer)
        {
            Debug.Assert(connectedLayer != null);

            ConnectedLayer = connectedLayer;
            Type = type;
        }

        List<Tuple<FlowDirection, Layer>> otherLayers = new List<Tuple<FlowDirection, Layer>>();

        bool suppressOtherSideUpdate;

        internal ConnectionType Type { get; private set; }

        internal Layer ConnectedLayer { get; private set; }

        public int Count
        {
            get { return otherLayers.Count; }
        }

        public IEnumerable<Layer> GetConnectedLayers(FlowDirection direction)
        {
            return otherLayers.Where(i => (int)(i.Item1 & direction) != 0).Select(i => i.Item2);
        }

        public void AddOneWay(Layer item)
        {
            Add(item, FlowDirection.OneWay);
        }

        public void AddTwoWay(Layer item)
        {
            Add(item, FlowDirection.TwoWay);
        }

        public void AddOneWayToSource(Layer item)
        {
            Add(item, FlowDirection.OneWayToSource);
        }

        public void Add(Layer item, FlowDirection direction)
        {
            Args.Requires(() => item, () => item != null);
            Args.Requires(() => direction, () => direction == FlowDirection.OneWay || direction == FlowDirection.OneWayToSource || direction == FlowDirection.TwoWay);

            if (!otherLayers.Any(l => l.Item2 == item))
            {
                WithOtherSideUpdateSuppressed(item, () =>
                {
                    if (Type == ConnectionType.Output)
                    {
                        item.InputConnections.Add(ConnectedLayer, direction);
                    }
                    else
                    {
                        item.OutputConnections.Add(ConnectedLayer, direction);
                    }
                });
                otherLayers.Add(Tuple.Create(direction, item));
            }
        }

        public bool Remove(Layer item)
        {
            Args.Requires(() => item, () => item != null);

            var toRemove = otherLayers.FirstOrDefault(i => i.Item2 == item);
            if (toRemove != null && otherLayers.Remove(toRemove))
            {
                WithOtherSideUpdateSuppressed(item, () =>
                {
                    if (Type == ConnectionType.Output)
                    {
                        item.InputConnections.Remove(ConnectedLayer);
                    }
                    else
                    {
                        item.OutputConnections.Remove(ConnectedLayer);
                    }
                });
                return true;
            }
            return false;
        }

        public void Clear()
        {
            foreach (var layer in otherLayers.Select(i => i.Item2))
            {
                if (Type == ConnectionType.Output)
                {
                    layer.InputConnections.Remove(ConnectedLayer);
                }
                else
                {
                    layer.OutputConnections.Remove(ConnectedLayer);
                }
            }
            otherLayers.Clear();
        }

        private void WithOtherSideUpdateSuppressed(Layer item, Action method)
        {
            if (!suppressOtherSideUpdate)
            {
                ((LayerConnections)item.InputConnections).suppressOtherSideUpdate = true;
                ((LayerConnections)item.OutputConnections).suppressOtherSideUpdate = true;
                try
                {
                    method();
                }
                finally
                {
                    ((LayerConnections)item.InputConnections).suppressOtherSideUpdate = false;
                    ((LayerConnections)item.OutputConnections).suppressOtherSideUpdate = false;
                }
            }
        }
    }
}
